### DISCUSSION TITLE
[node] process.exitCode is readonly

### DISCUSSION THREAD

**Discussion by txk2048** [DefinitelyTyped/DefinitelyTyped#57677] [Issues with a @types package]
Small example code:

```typescript
import * as fs from "fs";
import * as process from "process";

function main() {
  const args = process.argv.slice(2);
  if (args.length != 1) {
    console.error("You must provide a day to run");
    process.exitCode = 1;
    return;
  }

  console.log("abc");
}

main();
```

Trying to write a value to process.exitCode produces the following error message:

`error TS2540: Cannot assign to 'exitCode' because it is a read-only property.`

Having checked the definitions file I do not understand why I am getting this warning when the exitCode property does not appear to be set as readonly.

Package versions:

| Package          |  Version |
| :--------------: | :--------: |
| Node              | 16.9.1     |
| TypeScript      | 4.5.3       |
| @types/node | 16.11.2   |

[a comment by an automated account is not shown]

**Comment by peterblazejewicz** [comment]
@txk64 it comes from 'process' module, which has `exitCode` specified as `readonly`. I can see this can be set at runtime.

**Reply by txk2048** [reply]
I do not see the `readonly` keyword in front of the property declaration, so how does it end up being marked as `readonly`?

[Link to line in process definition file](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/node/process.d.ts#L627)

**Comment by txk2048** [comment]
I have just noticed some strange behaviour regarding the process module and the type definitions.

- If you do not explicitly import the process module and instead rely on the implicitly available object at runtime, then the compiler doesn't complain about `exitCode` being `readonly`.

- If you import the module however, the compiler then complains the property is `readonly`.

This feels like a bug with the definitions more so because in the Node.js docs they recommend importing it explicitly.
"_While it is available as a global, it is recommended to explicitly access it via require or import:_" [link to docs at location of quote](https://nodejs.org/api/process.html#process)

**Comment by SimonSchick** [comment]
When using `import * as thing` all properties of `thing` are readonly, this is not a DT bug, this is incorrect usage of imports.

**Reply by txk2048** [reply]
Does this mean the advice in the Node.js docs to explicitly import `process` should be avoided in TS?

**Reply by SimonSchick** [reply]
If you use `esModuleInterop` you should be able to `import process from 'process';` and manipulate the exit code as you desire since it's no longer considered an import proxy object (you are using `import * as process`

**Reply by flevi29** [reply]
Just use `import { default as process } from "node:process";`.

I don't like `esModuleInterop`, it's so unnecessary just for a minimal convenience, please don't use it if you don't absolutely have to. It barfs out quite a bit of useless code otherwise. The only time you might have to use it is if you're type checking an external library, but even then it's better to just [`skipLibCheck`](https://www.typescriptlang.org/tsconfig#skipLibCheck), it's even recommended by TypeScript.
